src/ast_editor_state.ml

```diff
@@ -0,0 +1,104 @@
+open Import
+
+type t =
+  { mutable webview_map : (string, WebView.t, String.comparator_witness) Map.t
+  ; mutable original_mode : bool
+  ; mutable hover_disposable : Disposable.t option
+  ; mutable pp_doc_to_changed_origin_map :
+      (string, bool, String.comparator_witness) Map.t
+  ; mutable origin_to_pp_doc_map :
+      (string, string, String.comparator_witness) Map.t
+  }
+
+let make () =
+  let webview_map = Map.empty (module String) in
+  let original_mode = true in
+  let hover_disposable = None in
+  let pp_doc_to_changed_origin_map = Map.empty (module String) in
+  let origin_to_pp_doc_map = Map.empty (module String) in
+  { webview_map
+  ; original_mode
+  ; hover_disposable
+  ; pp_doc_to_changed_origin_map
+  ; origin_to_pp_doc_map
+  }
+
+let find_original_doc_by_pp_uri ~uri_string map =
+  let r =
+    Map.filteri ~f:(fun ~key:_ ~data -> String.equal uri_string data) map
+    |> Map.to_alist |> List.unzip |> fst
+  in
+  match r with
+  | k :: _ -> Some k
+  | _ -> None
+
+let find_webview_by_doc t ~document =
+  let doc_uri = Uri.toString (TextDocument.uri document) () in
```

Comment:
Should be `to_string`.

Reply:
Do you mean changing the name of https://github.com/microsoft/vscode/blob/c3f126316369cd610563c75b1b1725e0679adfb3/src/vs/vscode.d.ts#L1413 ?

Reply:
Oh sorry - never mind. I thought Uri was our own module.